src/io/stm_tests.ml

```diff
@@ -0,0 +1,298 @@
+             3,return Pos;
+             3,return Length;
+             1,return Close;
+             1,return Close_noerr;
+             3,return Flush;
+             3,map (fun c -> Output_char c) char_gen;
+             3,map (fun i -> Output_byte i) byte_gen;
+             3,map (fun c -> Output_string c) string_gen;
+             3,map (fun b -> Output_bytes b) bytes_gen;
+             3,map3 (fun b p l -> Output (b,p,l)) bytes_gen byte_gen byte_gen;
+             3,map3 (fun s p l -> Output_substring (s,p,l)) string_gen byte_gen byte_gen;
+             3,map (fun b -> Set_binary_mode b) Gen.bool;
+             3,map (fun b -> Set_buffered b) Gen.bool;
+             3,return Is_buffered;
+           ]))
+
+  let init_state  = Closed 0L
+
+  let count_nls s =
+    String.fold_right (fun c count -> if c = '\n' then 1+count else count) s 0
+
+  let next_state c s = match c,s with
+    | Open_text, Closed _l ->
+      Open { position = 0L;
+             length = 0L;
+             buffered = true;
+             binary_mode = false; }
+    | Open_text, Open _ -> s
+    (* non-open cmd on closed Out_channel *)
+    | Output_char _, Closed _
+    | Output_byte _, Closed _
+    | Output_string _, Closed _
+    | Output_bytes _, Closed _
+    | Output (_,_,_), Closed _
+    | Output_substring (_,_,_), Closed _
+    | Seek _, Closed _
+    | Close, Closed _
+    | Close_noerr, Closed _
+    | Set_binary_mode _, Closed _
+    | Set_buffered _, Closed _ -> s
+    (* non-open cmd on open Out_channel *)
+    | Seek p, Open { position = _; length; buffered; binary_mode } ->
+      Open { position = p;
+             length = Int64.max length p;
+             buffered;
+             binary_mode; }
+    | Pos,_ -> s
+    | Length,_ -> s
+    | Close, Open _ -> Closed 0L
+    | Close_noerr, Open _ -> Closed 0L
+    | Flush, _ -> s
+    | Set_binary_mode b, Open { position; length; buffered; binary_mode = _ } ->
+      Open { position; length; buffered; binary_mode = b }
+    | Is_buffered, _ -> s
+    | Set_buffered b, Open { position; length; buffered = _; binary_mode } ->
+      Open { position; length; buffered = b; binary_mode }
+    (* output on open Out_channel *)
+    | Output_char c, Open { position; length; buffered; binary_mode } ->
+      let position = Int64.succ position in
+      let length = (* Windows text mode maps '\n' to "\r\n" *)
+        Int64.add length
+          (if (Sys.win32 || Sys.cygwin) && not binary_mode && c = '\n' then 2L else 1L) in
+      Open { position; length; buffered; binary_mode; }
+    | Output_byte i, Open { position; length; buffered; binary_mode } ->
+      let position = Int64.succ position in
+      let length = (* Windows text mode maps '\n' to "\r\n" *)
+        Int64.add length
+          (if (Sys.win32 || Sys.cygwin) && not binary_mode && (i mod 256 = 10) then 2L else 1L) in
+      Open { position; length; buffered; binary_mode; }
+    | Output_string arg, Open { position; length; buffered; binary_mode } ->
```

Comment:
I wondered there and on the following cases whether it could be possible to factor most of the logic out. I didn’t see an obviously good way to go, though.

Reply:
There is indeed a lot of repetition!
I suggest a simple refactoring in 9bd3253